contribute/design/1.x/cstor/zcmd-operation.md

```diff
@@ -0,0 +1,232 @@
+	VersionDetails VersionDetails  `json:"versionDetails"`
+}
+
+// CStorPoolSpec is the spec listing fields for a CStorPool resource.
+type CStorPoolSpec struct {
+	Group    []BlockDeviceGroup `json:"group"`
+	PoolSpec CStorPoolAttr      `json:"poolSpec"`
+    // CommandExecution will contain the required information to execute the command
+    CommandExecution CommandExecution
+}
+
+// CommandExecution contains the required details to execute command
+type CommandExecution struct {
+    // Command that need to execute on corresponding cStor pool
+    Command string
+    // MaxRetryCount desribes the max retries controller try for execution
+    MaxRetryCount int
+    // CurreentRetryCount describes the current iteration controller executing the command
+    CurrentRetryCount int
+}
+
+// BlockDeviceGroup contains a collection of block device for a given pool topology in CSP.
+type BlockDeviceGroup struct {
+	// Item contains a list of CspBlockDevice.
+	Item []CspBlockDevice `json:"blockDevice"`
+}
+
+// CspBlockDevice contains the details of block device present on CSP.
+type CspBlockDevice struct {
+	// Name is the name of the block device resource.
+	Name string `json:"name"`
+	// DeviceID is the device id of the block device resource. In case of sparse
+	// block device, it contains the device path.
+	DeviceID string `json:"deviceID"`
+	// InUseByPool tells whether the block device is present on spc. If block
+	// device is present on SPC, it is true else false.
+	InUseByPool bool `json:"inUseByPool"`
+}
+
+// BlockDeviceAttr stores the block device related attributes.
+type BlockDeviceAttr struct {
+	BlockDeviceList []string `json:"blockDeviceList"`
+}
+
+// CStorPoolAttr is to describe zpool related attributes.
+type CStorPoolAttr struct {
+	CacheFile string `json:"cacheFile"` //optional, faster if specified
+	PoolType  string `json:"poolType"`  //mirrored, striped
+	// OverProvisioning field is deprecated and not honoured
+	OverProvisioning bool `json:"overProvisioning"` //true or false
+	// ThickProvisioning, If true disables OverProvisioning
+	ThickProvisioning bool `json:"thickProvisioning"` // true or false
+	// ROThresholdLimit is threshold(percentage base) limit for pool read only mode, if ROThresholdLimit(%) of pool storage is used then pool will become readonly, CVR also. (0 < ROThresholdLimit < 100, default:100)
+	ROThresholdLimit int `json:"roThresholdLimit"` //optional
+}
+
+// CStorPoolStatus is for handling status of pool.
+type CStorPoolStatus struct {
+	Phase    CStorPoolPhase        `json:"phase"`
+	Capacity CStorPoolCapacityAttr `json:"capacity"`
+
+	// LastTransitionTime refers to the time when the phase changes
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+	//ReadOnly if pool is readOnly or not
+	ReadOnly bool `json:"readOnly"`
+
+    // CommandResult contains the output of command requested in spec section
+    CommandResult string `json:"commandResult"`
```

Comment:
Few questions
 - what is the maximum limit for output?
 - if a command with trailing output is provided what will be the behaviour? Will the truncated output come here or it wait for the entire output?

Reply:
> what is the maximum limit for output?

As JSON string doesn't contain any limits AFAIK... but when unmarshaled into go struct it will depend on pod memory limits...

> if a command with trailing output is provided what will be the behaviour? Will the truncated output come here or it wait for the entire output?

Can you elaborate on what is meant by trailing output?

Reply:
> Can you elaborate on what is meant by trailing output?

If a command takes more time to  execute than the timeout of the sync then the output will come as truncated or it will not come?

Reply:
It will be an atomic operation upon failure/success only we will update the output

Reply:
Then it is good. What if we want to execute that takes more time than the sync period? Do we need to change the sync period as well?

Reply:
Actually there is maxRetryCount(new filed) controller will retry the command for specified no.of times and the sync period can have any value and high-level operators shouldn't worry about the sync period.

Q: I am thinking to add another field in status to represent whether the command succeeded/failed. WDYT?

Reply:
Yes makes sense, we cannot assume empty output as a failure.